src/mavsdk/plugins/ftp/ftp_impl.cpp

```diff
@@ -1306,12 +1309,99 @@ FtpImpl::ServerResult FtpImpl::_work_calc_file_CRC32(PayloadHeader* payload)
     return ServerResult::SUCCESS;
 }
 
+unsigned FtpImpl::get_size()
+{
+    if (_session_info.stream_download) {
+        return MAVLINK_MSG_ID_FILE_TRANSFER_PROTOCOL_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES;
+
+    } else {
+        return 0;
+    }
+}
+
 void FtpImpl::send()
 {
     // Anything to stream?
     if (!_session_info.stream_download) {
         return;
     }
+
+    unsigned max_bytes_to_send = 20000; // _mavlink->get_free_tx_buf();
+
+    // Send stream packets until buffer is full
+    bool more_data = false;
+
+    ServerResult error_code = ServerResult::SUCCESS;
+    mavlink_file_transfer_protocol_t ftp_msg;
+    PayloadHeader* payload = reinterpret_cast<PayloadHeader*>(&ftp_msg.payload[0]);
+    payload->seq_number = _session_info.stream_seq_number;
+    payload->session = 0;
+    payload->opcode = RSP_ACK;
+    payload->req_opcode = CMD_BURST_READ_FILE;
+    payload->offset = _session_info.stream_offset;
+    _session_info.stream_seq_number++;
+    // We have to test seek past EOF ourselves, lseek will allow seek past EOF
+    if (_session_info.stream_offset >= _session_info.file_size) {
+        error_code = ServerResult::ERR_EOF;
+    }
+
+    if (error_code == ServerResult::SUCCESS) {
+        if (lseek(_session_info.fd, payload->offset, SEEK_SET) < 0) {
+            error_code = ServerResult::ERR_FAIL;
+        }
+    }
+
+    if (error_code == ServerResult::SUCCESS) {
+        int bytes_read = ::read(_session_info.fd, &payload->data[0], max_data_length);
+
+        if (bytes_read < 0) {
+            // Negative return indicates error other than eof
+            error_code = ServerResult::ERR_FAIL;
+        } else {
+            payload->size = bytes_read;
+            _session_info.stream_offset += bytes_read;
+            _session_info.stream_chunk_transmitted += bytes_read;
+            LogDebug() << "Burst step: " << _session_info.stream_offset << " < "
+                       << _session_info.file_size;
+            more_data = (_session_info.stream_offset < _session_info.file_size);
+            if (!more_data) {
+                LogDebug() << "Burst completed";
+                payload->burst_complete = true;
+                _session_info.stream_download = false;
+                _session_info.stream_chunk_transmitted = 0;
+
+            } else {
+                more_data = true;
+                payload->burst_complete = false;
+                max_bytes_to_send -= get_size();
+            }
+        }
+    }
+
+    if (error_code != ServerResult::SUCCESS) {
+        payload->opcode = RSP_NAK;
+        payload->size = 1;
+        uint8_t* pData = &payload->data[0];
+        *pData =
+            error_code; // Straight reference to data[0] is causing bogus gcc array subscript error
```

Comment:
What?

Reply:
The whole mavsdk ftp plugin seems to be based on the PX4 implementation. So I based the ftp burst operation also on the PX4 implementation (this bogus comment comes from PX4: https://github.com/PX4/PX4-Autopilot/blob/f8c2ee73db3654a34976291e5ac94c437e27d4a7/src/modules/mavlink/mavlink_ftp.cpp#L1126)

The implementation here is not complete yet; I transmit the entire file within one burst. Appearently, in PX4, they burst out up to 35000 bytes at once, but it's not specified yet: https://mavlink.io/en/services/ftp.html#burst-read-file

Reply:
Aaaah, good point. We should do that...